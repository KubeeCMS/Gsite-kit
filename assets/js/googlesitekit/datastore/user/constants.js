/**
 * core/user data store: constants.
 *
 * Site Kit by Google, Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const STORE_NAME = 'core/user';

// Permissions list.
export const PERMISSION_AUTHENTICATE = 'googlesitekit_authenticate';
export const PERMISSION_SETUP = 'googlesitekit_setup';
export const PERMISSION_VIEW_POSTS_INSIGHTS = 'googlesitekit_view_posts_insights';
export const PERMISSION_VIEW_DASHBOARD = 'googlesitekit_view_dashboard';
export const PERMISSION_VIEW_MODULE_DETAILS = 'googlesitekit_view_module_details';
export const PERMISSION_MANAGE_OPTIONS = 'googlesitekit_manage_options';
export const PERMISSION_PUBLISH_POSTS = 'googlesitekit_publish_posts';

// Error code returned when scopes are missing.
export const ERROR_MISSING_REQUIRED_SCOPE = 'missing_required_scopes';
