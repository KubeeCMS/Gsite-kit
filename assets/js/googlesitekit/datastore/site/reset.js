/**
 * core/site data store: reset connection.
 *
 * Site Kit by Google, Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import API from 'googlesitekit-api';
import Data from 'googlesitekit-data';
import { STORE_NAME } from './constants';
import { createFetchStore } from '../../data/create-fetch-store';

const { createRegistrySelector } = Data;

const fetchResetStore = createFetchStore( {
	baseName: 'reset',
	controlCallback: () => {
		return API.set( 'core', 'site', 'reset' );
	},
	reducerCallback: () => {
		return {
			...INITIAL_STATE,
		};
	},
} );

const BASE_INITIAL_STATE = {};

const baseActions = {
	/**
	 * Resets the website's connection info to Site Kit.
	 *
	 * WARNING: This causes the website's connection with Google Site Kit to be
	 * removed and will require re-authentication. Use this action with caution,
	 * and always request user confirmation before dispatching.
	 *
	 * @since 1.5.0
	 */
	*reset() {
		yield fetchResetStore.actions.fetchReset();
	},
};

const baseSelectors = {
	/**
	 * Checks if reset action is in-process.
	 *
	 * @since 1.5.0
	 *
	 * @return {boolean} `true` if resetting is in-flight; `false` if not.
	 */
	isDoingReset: createRegistrySelector( ( select ) => () => {
		return select( STORE_NAME ).isFetchingReset();
	} ),
};

const store = Data.combineStores(
	fetchResetStore,
	{
		INITIAL_STATE: BASE_INITIAL_STATE,
		actions: baseActions,
		selectors: baseSelectors,
	}
);

export const INITIAL_STATE = store.INITIAL_STATE;
export const actions = store.actions;
export const controls = store.controls;
export const reducer = store.reducer;
export const resolvers = store.resolvers;
export const selectors = store.selectors;

export default store;
