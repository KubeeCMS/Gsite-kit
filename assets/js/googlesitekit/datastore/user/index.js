/**
 * core/user data store
 *
 * Site Kit by Google, Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import Data from 'googlesitekit-data';
import authentication from './authentication';
import dateRange from './date-range';
import disconnect from './disconnect';
import error from './error';
import notifications from './notifications';
import permissions from './permissions';
import userInfo from './user-info';
import { STORE_NAME } from './constants';

export { STORE_NAME };

const store = Data.combineStores(
	Data.commonStore,
	authentication,
	dateRange,
	disconnect,
	error,
	notifications,
	permissions,
	userInfo,
);

export const {
	INITIAL_STATE,
	actions,
	controls,
	reducer,
	resolvers,
	selectors,
} = store;

// Register this store on the global registry.
Data.registerStore( STORE_NAME, store );

export default store;
