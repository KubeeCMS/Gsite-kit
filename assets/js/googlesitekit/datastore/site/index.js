/**
 * core/site data store
 *
 * Site Kit by Google, Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import Data from 'googlesitekit-data';
import connection from './connection';
import html from './html';
import info from './info';
import reset from './reset';
import { STORE_NAME } from './constants';
import notifications from './notifications';
import registryKey from './registry-key';

export { STORE_NAME };

const store = Data.combineStores(
	Data.commonStore,
	connection,
	html,
	info,
	reset,
	notifications,
	registryKey
);

export const INITIAL_STATE = store.INITIAL_STATE;
export const actions = store.actions;
export const controls = store.controls;
export const reducer = store.reducer;
export const resolvers = store.resolvers;
export const selectors = store.selectors;

// Register this store on the global registry.
Data.registerStore( STORE_NAME, store );

export default store;
