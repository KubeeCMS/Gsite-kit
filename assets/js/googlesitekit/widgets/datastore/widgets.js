/**
 * core/widgets data store: widgets info.
 *
 * Site Kit by Google, Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import invariant from 'invariant';
import { v4 as uuidv4 } from 'uuid';

/**
 * Internal dependencies
 */
import Data from 'googlesitekit-data';
import { WIDGET_WIDTHS } from './constants';
import { sortByProperty } from '../../../util/sort-by-property';
import { STORE_NAME as CORE_SITE } from '../../../googlesitekit/datastore/site/constants';

const { commonActions, createRegistrySelector } = Data;

/**
 * Store our widget components by registry, then by widget `slug`. We do this because
 * we can't store React components in our data store.
 *
 * @private
 * @since 1.9.0
 */
export const WidgetComponents = {};

const ASSIGN_WIDGET = 'ASSIGN_WIDGET';
const REGISTER_WIDGET = 'REGISTER_WIDGET';

const WidgetWidthKeys = Object.keys( WIDGET_WIDTHS ).map( ( ( key ) => `WIDGET_WIDTHS.${ key }` ) ).join( ', ' );

export const INITIAL_STATE = {
	areaAssignments: {},
	registryKey: undefined,
	widgets: {},
};

export const actions = {
	/**
	 * Assigns an existing widget (by slug) to a widget area(s).
	 *
	 * @since 1.9.0
	 *
	 * @param {string}         slug      Widget slug.
	 * @param {(string|Array)} areaSlugs Widget Area slug(s).
	 * @return {Object} Redux-style action.
	 */
	assignWidget( slug, areaSlugs ) {
		return {
			payload: {
				slug,
				areaSlugs: ( typeof areaSlugs === 'string' ) ? [ areaSlugs ] : areaSlugs,
			},
			type: ASSIGN_WIDGET,
		};
	},

	/**
	 * Register a widget with a given slug and settings.
	 *
	 * @since 1.9.0
	 * @since 1.12.0 Added wrapWidget setting.
	 *
	 * @param {string}      slug                Widget's slug.
	 * @param {Object}      settings            Widget's settings.
	 * @param {WPComponent} settings.component  React component used to display the contents of this widget.
	 * @param {number}      settings.priority   Optional. Widget's priority for ordering (lower number is higher priority, like WordPress hooks). Default is: 10.
	 * @param {string}      settings.width      Optional. Widget's maximum width to occupy. Default is: "quarter". One of: "quarter", "half", "full".
	 * @param {boolean}     settings.wrapWidget Optional. Whether to wrap the component with the <Widget> wrapper. Default is: true.
	 * @return {Object} Redux-style action.
	 */
	*registerWidget( slug, {
		component,
		priority = 10,
		width = WIDGET_WIDTHS.QUARTER,
		wrapWidget = true,
	} = {} ) {
		invariant( component, 'component is required to register a widget.' );
		invariant( Object.values( WIDGET_WIDTHS ).includes( width ), `Widget width should be one of: ${ WidgetWidthKeys }, but "${ width }" was provided.` );

		const registry = yield commonActions.getRegistry();
		let registryKey = yield registry.select( CORE_SITE ).getRegistryKey();

		if ( registryKey === undefined ) {
			registryKey = uuidv4();
			yield registry.dispatch( CORE_SITE ).setRegistryKey( registryKey );
		}

		// We do this assignment in the action rather than the reducer because we can't send a
		// payload that includes a React component to the reducer; we'll get an error about
		// payloads needing to be plain objects.
		if ( WidgetComponents[ registryKey ] === undefined ) {
			WidgetComponents[ registryKey ] = {};
		}
		if ( WidgetComponents[ registryKey ][ slug ] === undefined ) {
			WidgetComponents[ registryKey ][ slug ] = component;
		}

		yield {
			payload: { slug, settings: { priority, width, wrapWidget } },
			type: REGISTER_WIDGET,
		};

		return {};
	},
};

export const controls = {};

export const reducer = ( state, { type, payload } ) => {
	switch ( type ) {
		case ASSIGN_WIDGET: {
			const { slug, areaSlugs } = payload;

			const { areaAssignments } = state;
			areaSlugs.forEach( ( areaSlug ) => {
				if ( areaAssignments[ areaSlug ] === undefined ) {
					areaAssignments[ areaSlug ] = [];
				}

				if ( ! areaAssignments[ areaSlug ].includes( slug ) ) {
					areaAssignments[ areaSlug ].push( slug );
				}
			} );

			return {
				...state,
				areaAssignments,
			};
		}

		case REGISTER_WIDGET: {
			const { slug, settings } = payload;

			if ( state.widgets[ slug ] !== undefined ) {
				global.console.warn( `Could not register widget with slug "${ slug }". Widget "${ slug }" is already registered.` );

				return { ...state };
			}

			return {
				...state,
				widgets: {
					...state.widgets,
					[ slug ]: { ...settings, slug },
				},
			};
		}

		default: {
			return { ...state };
		}
	}
};

export const resolvers = {
};

export const selectors = {
	/**
	 * Checks if a widget has been registered with a given slug.
	 *
	 * Returns `true` if the widget area has been registered.
	 * Returns `false` if the widget area has NOT been registered.
	 *
	 * @since 1.9.0
	 *
	 * @param {Object} state Data store's state.
	 * @param {string} slug  Widget's slug.
	 * @return {boolean} `true`/`false` based on whether widget has been registered.
	 */
	isWidgetRegistered( state, slug ) {
		const { widgets } = state;

		return widgets[ slug ] !== undefined;
	},

	/**
	 * Returns all widgets registered for a given widget area.
	 *
	 * Returns an array of all widgets for a given area.
	 * The widgets are returned in order of their priority, so can be rendered in
	 * the order provided by the selector.
	 *
	 * @since 1.9.0
	 *
	 * @param {Object} state          Data store's state.
	 * @param {string} widgetAreaSlug Widget context to get areas for.
	 * @return {Array} An ordered array of widgets for this area.
	 */
	getWidgets: createRegistrySelector( ( select ) => ( state, widgetAreaSlug ) => {
		invariant( widgetAreaSlug, 'widgetAreaSlug is required.' );

		const { areaAssignments, widgets } = state;

		const registryKey = select( CORE_SITE ).getRegistryKey();

		const sorted = sortByProperty(
			Object.values( widgets ).filter( ( widget ) => {
				return areaAssignments[ widgetAreaSlug ] && areaAssignments[ widgetAreaSlug ].includes( widget.slug );
			} ).map( ( widget ) => {
				const widgetWithComponent = { ...widget };
				if ( WidgetComponents[ registryKey ] ) {
					widgetWithComponent.component = WidgetComponents[ registryKey ][ widget.slug ];
				}

				return widgetWithComponent;
			} ),
			'priority'
		);
		return sorted;
	} ),

	/**
	 * Returns a single widget, by slug.
	 *
	 * @since 1.11.0
	 *
	 * @param {Object} state  Data store's state.
	 * @param {string} slug   Widget slug.
	 * @return {Object|null} A widget object, if one exists.
	 */
	getWidget: createRegistrySelector( ( select ) => ( state, slug ) => {
		invariant( slug, 'slug is required to get a widget.' );

		const { widgets } = state;

		const registryKey = select( CORE_SITE ).getRegistryKey();

		const widget = widgets[ slug ];
		if ( widget && WidgetComponents[ registryKey ] ) {
			widget.component = WidgetComponents[ registryKey ][ widget.slug ];
		}

		return widget || null;
	} ),
};

export default {
	INITIAL_STATE,
	actions,
	controls,
	reducer,
	resolvers,
	selectors,
};
