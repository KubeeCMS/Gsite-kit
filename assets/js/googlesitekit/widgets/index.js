/**
 * Widgets API
 *
 * Site Kit by Google, Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { React } from '@wordpress/element';

/**
 * Internal dependencies
 */
import { dispatch, select } from 'googlesitekit-data';
import Widget from './components/Widget';
import { STORE_NAME, WIDGET_WIDTHS, WIDGET_STYLES } from './datastore/constants';
// This import has side-effects; it registers the Widgets datastore on the default
// data store registry (eg. `googlesitekit.data`).
import './datastore';

export { registerDefaults } from './register-defaults';

const Widgets = {
	/**
	 * Public Widget components for creating Site Kit widgets.
	 *
	 * @since 1.11.0
	 */
	components: { Widget },

	/**
	 * Supported styles for Site Kit widgets.
	 *
	 * @since n.e.x.t
	 */
	WIDGET_STYLES,

	/**
	 * Supported widths for Site Kit widgets.
	 *
	 * @since n.e.x.t
	 */
	WIDGET_WIDTHS,

	/**
	 * Registers a widget area.
	 *
	 * @since 1.9.0
	 *
	 * @param {string}             slug                Widget area's slug.
	 * @param {Object}             settings            Widget area's settings.
	 * @param {string}             settings.title      Title for this widget area.
	 * @param {string}             settings.subtitle   Subtitle for this widget area.
	 * @param {number}             [settings.priority] Optional. Priority for this widget area. Default: 10.
	 * @param {(string|undefined)} [settings.icon]     Optional. URL to SVG icon for this widget area.
	 * @param {string}             [settings.style]    Optional. Widget area style (one of "boxes", "composite"). Default: "boxes".
	 * @param {(string|Array)}     [contextSlugs]      Optional. Widget context slug(s).
	 */
	registerWidgetArea( slug, settings, contextSlugs ) {
		dispatch( STORE_NAME ).registerWidgetArea( slug, settings );
		if ( contextSlugs ) {
			Widgets.assignWidgetArea( slug, contextSlugs );
		}
	},

	/**
	 * Registers a widget.
	 *
	 * @since 1.9.0
	 *
	 * @param {string}          slug                Widget's slug.
	 * @param {Object}          settings            Widget's settings.
	 * @param {React.Component} settings.component  React component used to display the contents of this widget.
	 * @param {number}          settings.priority   Optional. Widget's priority for ordering (lower number is higher priority, like WordPress hooks). Default is: 10.
	 * @param {string}          settings.width      Optional. Widget's maximum width to occupy. Default is: "quarter". One of: "quarter", "half", "full".
	 * @param {boolean}         settings.wrapWidget Optional. Whether to wrap the component with the <Widget> wrapper. Default is: true.
	 * @param {(string|Array)}  [widgetAreaSlugs]   Optional. Widget area slug(s).
	 */
	registerWidget( slug, settings, widgetAreaSlugs ) {
		dispatch( STORE_NAME ).registerWidget( slug, settings );
		if ( widgetAreaSlugs ) {
			Widgets.assignWidget( slug, widgetAreaSlugs );
		}
	},

	/**
	 * Assigns a widget area to one (or several) contexts.
	 *
	 * Accepts an area slug to register as the first argument, then either a string
	 * (for a single context slug) or array of contexts slugs (to assign the widget area
	 * to multiple contexts).
	 *
	 * @since 1.9.0
	 *
	 * @param {string}         slug          Widget Area's slug.
	 * @param {(string|Array)} contextSlugs  Widget context slug(s).
	 */
	assignWidgetArea( slug, contextSlugs ) {
		dispatch( STORE_NAME ).assignWidgetArea( slug, contextSlugs );
	},

	/**
	 * Assigns an existing widget (by slug) to a widget area(s).
	 *
	 * @since 1.9.0
	 *
	 * @param  {string}         slug            Widget slug.
	 * @param  {(string|Array)} widgetAreaSlugs Widget area slug(s).
	 */
	assignWidget( slug, widgetAreaSlugs ) {
		dispatch( STORE_NAME ).assignWidget( slug, widgetAreaSlugs );
	},

	/**
	 * Checks if a widget area has been registered.
	 *
	 * Returns `true` if the widget area has been registered.
	 * Returns `false` if the widget area has NOT been registered.
	 *
	 * @since 1.9.0
	 *
	 * @param  {string} slug Widget Area's slug.
	 * @return {boolean} `true`/`false` based on whether widget area has been registered.
	 */
	isWidgetAreaRegistered( slug ) {
		return select( STORE_NAME ).isWidgetAreaRegistered( slug );
	},

	/**
	 * Checks if a widget has been registered with a given slug.
	 *
	 * Returns `true` if the widget area has been registered.
	 * Returns `false` if the widget area has NOT been registered.
	 *
	 * @since 1.9.0
	 *
	 * @param  {string}  slug  Widget's slug.
	 * @return {boolean} `true`/`false` based on whether widget has been registered.
	 */
	isWidgetRegistered( slug ) {
		return select( STORE_NAME ).isWidgetRegistered( slug );
	},
};

export default Widgets;
