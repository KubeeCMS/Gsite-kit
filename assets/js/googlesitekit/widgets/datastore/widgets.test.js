/**
 * core/widgets data store: widget tests.
 *
 * Site Kit by Google, Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import {
	createTestRegistry,
	unsubscribeFromAll,
	muteConsole,
} from '../../../../../tests/js/utils';
import { render } from '../../../../../tests/js/test-utils';
import { STORE_NAME } from './constants';
import { STORE_NAME as CORE_SITE_STORE_NAME } from '../../../googlesitekit/datastore/site';
import { WidgetComponents } from './widgets';

describe( 'core/widgets Widgets', () => {
	const resetWidgetComponents = () => {
		Object.keys( WidgetComponents ).forEach( ( registryKey ) => {
			delete WidgetComponents[ registryKey ];
		} );
	};
	let registry;
	let store;

	beforeEach( () => {
		registry = createTestRegistry();
		store = registry.stores[ STORE_NAME ].store;

		// Reset the WidgetComponents variable for each test; otherwise a new registry will
		// be created for each test.
		resetWidgetComponents();
	} );

	afterEach( () => {
		unsubscribeFromAll( registry );
	} );

	describe( 'actions', () => {
		describe( 'assignWidget', () => {
			it( 'should assign widgets to a widget area', () => {
				const WidgetComponent = () => {
					return ( <div>Foo bar!</div> );
				};
				const AnotherWidgetComponent = () => {
					return ( <div>Howdy friend!</div> );
				};
				registry.dispatch( STORE_NAME ).registerWidgetArea( 'dashboard-header', {
					title: 'Dashboard Header',
					subtitle: 'Cool stuff for yoursite.com',
					style: 'boxes',
				} );
				registry.dispatch( STORE_NAME ).assignWidgetArea( 'dashboard-header', 'dashboard' );

				// Register all the widgets.
				const slugOne = 'one';
				const settingsOne = {
					component: WidgetComponent,
					priority: 5,
				};
				registry.dispatch( STORE_NAME ).registerWidget( slugOne, settingsOne );

				const slugTwo = 'two';
				const settingsTwo = {
					component: WidgetComponent,
					priority: 10,
				};
				registry.dispatch( STORE_NAME ).registerWidget( slugTwo, settingsTwo );

				const slugThree = 'three';
				const settingsThree = {
					component: AnotherWidgetComponent,
					priority: 500,
				};
				registry.dispatch( STORE_NAME ).registerWidget( slugThree, settingsThree );

				// Assign all widgets to the widget area.
				registry.dispatch( STORE_NAME ).assignWidget( slugOne, 'dashboard-header' );
				registry.dispatch( STORE_NAME ).assignWidget( slugTwo, 'dashboard-header' );
				registry.dispatch( STORE_NAME ).assignWidget( slugThree, 'dashboard-header' );

				const widgets = registry.select( STORE_NAME ).getWidgets( 'dashboard-header' );

				expect( widgets ).toHaveLength( 3 );
				expect( widgets ).toMatchObject( [
					{ ...settingsOne, slug: slugOne },
					{ ...settingsTwo, slug: slugTwo },
					{ ...settingsThree, slug: slugThree },
				] );
			} );

			it( 'should allow assignment of non-registered widgets', () => {
				registry.dispatch( STORE_NAME ).registerWidgetArea( 'dashboard-header', {
					title: 'Dashboard Header',
					subtitle: 'Cool stuff for yoursite.com',
					style: 'boxes',
				} );
				registry.dispatch( STORE_NAME ).assignWidgetArea( 'dashboard-header', 'dashboard' );

				// Assign all widgets to the widget area.
				registry.dispatch( STORE_NAME ).assignWidget( 'slugOne', 'dashboard-header' );
				registry.dispatch( STORE_NAME ).assignWidget( 'slugTwo', 'dashboard-header' );
				registry.dispatch( STORE_NAME ).assignWidget( 'slugThree', 'dashboard-header' );

				const state = store.getState();

				expect( state.areaAssignments[ 'dashboard-header' ].includes( 'slugOne' ) ).toEqual( true );
				expect( state.areaAssignments[ 'dashboard-header' ].includes( 'slugTwo' ) ).toEqual( true );
				expect( state.areaAssignments[ 'dashboard-header' ].includes( 'slugThree' ) ).toEqual( true );
			} );

			it( 'should allow assignment of non-registered widget areas', () => {
				// Assign all widgets to the widget area.
				registry.dispatch( STORE_NAME ).assignWidget( 'testOne', 'dashboard-header' );
				registry.dispatch( STORE_NAME ).assignWidget( 'testTwo', 'dashboard-header' );

				const state = store.getState();

				expect( state.areaAssignments[ 'dashboard-header' ].includes( 'testOne' ) ).toEqual( true );
				expect( state.areaAssignments[ 'dashboard-header' ].includes( 'testTwo' ) ).toEqual( true );
			} );
		} );

		describe( 'registerWidget', () => {
			it( 'places the widget component in a variable external to the store, with a key from the store', () => {
				expect( Object.keys( WidgetComponents ) ).toHaveLength( 0 );

				const slug = 'WidgetSpinner';
				const WidgetSpinner = ( props ) => {
					return ( <div>Hello { props.children }!</div> );
				};
				registry.dispatch( STORE_NAME ).registerWidget( slug, {
					component: WidgetSpinner,
				} );
				const registryKey = registry.select( CORE_SITE_STORE_NAME ).getRegistryKey();

				// A registry key should have been created.
				expect( Object.keys( WidgetComponents ) ).toHaveLength( 1 );
				expect( WidgetComponents[ registryKey ][ slug ] ).toBeDefined();
				expect( WidgetComponents[ registryKey ][ slug ] ).toEqual( WidgetSpinner );

				// Ensure we can render a component with the widget's component, verifying it's still a
				// usable React component.
				const Component = WidgetComponents[ registryKey ][ slug ];
				const { container } = render( <Component>world</Component> );
				expect( container.firstChild ).toMatchSnapshot();
			} );

			it( 'uses the same WidgetsComponent section when multiple widgets are registered', () => {
				registry.dispatch( STORE_NAME ).registerWidget( 'widget-1', {
					component: () => {
						return ( <div>Hello world!</div> );
					},
				} );
				registry.dispatch( STORE_NAME ).registerWidget( 'widget-2', {
					component: () => {
						return ( <span>Hello again!</span> );
					},
				} );
				const registryKey = registry.select( CORE_SITE_STORE_NAME ).getRegistryKey();

				expect( Object.keys( WidgetComponents[ registryKey ] ) ).toHaveLength( 2 );
				// Another registry key should not have been created.
				expect( Object.keys( WidgetComponents ) ).toHaveLength( 1 );
			} );

			it( 'does not overwrite an existing widget', () => {
				const slug = 'widget-1';
				const WidgetOne = () => {
					return ( <div>Hello world!</div> );
				};
				const WidgetOneRedone = () => {
					return ( <div>Goodbye you!</div> );
				};
				registry.dispatch( STORE_NAME ).registerWidget( slug, {
					component: WidgetOne,
				} );

				// Mute warning about duplicate slug since we expect it below anyway.
				muteConsole( 'warn' );

				// Expect console warning about duplicate slug.
				const consoleWarnSpy = jest.spyOn( global.console, 'warn' );
				registry.dispatch( STORE_NAME ).registerWidget( slug, {
					component: WidgetOneRedone,
				} );
				expect( consoleWarnSpy ).toHaveBeenCalledWith( `Could not register widget with slug "${ slug }". Widget "${ slug }" is already registered.` );
				consoleWarnSpy.mockClear();

				const registryKey = registry.select( CORE_SITE_STORE_NAME ).getRegistryKey();

				// Ensure original widget's component is registered.
				expect( Object.keys( WidgetComponents[ registryKey ] ) ).toHaveLength( 1 );
				expect( WidgetComponents[ registryKey ][ slug ] ).toEqual( WidgetOne );
			} );
		} );
	} );

	describe( 'selectors', () => {
		describe( 'getWidgets', () => {
			it( 'requires a widgetAreaSlug', () => {
				expect( () => {
					registry.select( STORE_NAME ).getWidgets();
				} ).toThrow( 'widgetAreaSlug is required.' );
			} );

			it( 'should return an empty array when no widgets have been registered', () => {
				const widgets = registry.select( STORE_NAME ).getWidgets( 'dashboardHeader' );

				expect( widgets ).toEqual( [] );
			} );

			it( "should return all widgets for a given widgetAreaSlug after they're registered", () => {
				const PageViews = () => {
					return ( <div>Ten people viewed your page!</div> );
				};
				registry.dispatch( STORE_NAME ).registerWidgetArea( 'dashboard-header', {
					title: 'Dashboard Header',
					subtitle: 'Cool stuff for yoursite.com',
					style: 'boxes',
				} );
				registry.dispatch( STORE_NAME ).assignWidgetArea( 'dashboard-header', 'dashboard' );
				registry.dispatch( STORE_NAME ).registerWidget( 'PageViews', {
					component: PageViews,
				} );
				registry.dispatch( STORE_NAME ).assignWidget( 'PageViews', 'dashboard-header' );

				const widgets = registry.select( STORE_NAME ).getWidgets( 'dashboard-header' );

				expect( widgets ).toHaveLength( 1 );
			} );

			it( 'should return widgets in the correct priority', () => {
				const WidgetComponent = () => {
					return ( <div>Foo bar!</div> );
				};
				registry.dispatch( STORE_NAME ).registerWidgetArea( 'dashboard-header', {
					title: 'Dashboard Header',
					subtitle: 'Cool stuff for yoursite.com',
					style: 'boxes',
				} );
				registry.dispatch( STORE_NAME ).assignWidgetArea( 'dashboard-header', 'dashboard' );

				// Register all the widgets.
				registry.dispatch( STORE_NAME ).registerWidget( 'lowest', {
					component: WidgetComponent,
					priority: 5,
				} );
				registry.dispatch( STORE_NAME ).registerWidget( 'mediumOne', {
					component: WidgetComponent,
					priority: 10,
				} );
				registry.dispatch( STORE_NAME ).registerWidget( 'mediumTwo', {
					component: WidgetComponent,
					priority: 10,
				} );
				registry.dispatch( STORE_NAME ).registerWidget( 'highest', {
					component: WidgetComponent,
					priority: 500,
				} );

				// Assign all widgets to the widget area.
				registry.dispatch( STORE_NAME ).assignWidget( 'lowest', 'dashboard-header' );
				registry.dispatch( STORE_NAME ).assignWidget( 'mediumOne', 'dashboard-header' );
				registry.dispatch( STORE_NAME ).assignWidget( 'mediumTwo', 'dashboard-header' );
				registry.dispatch( STORE_NAME ).assignWidget( 'highest', 'dashboard-header' );

				const widgets = registry.select( STORE_NAME ).getWidgets( 'dashboard-header' );

				expect( widgets ).toHaveLength( 4 );
				expect( widgets[ 0 ].slug ).toEqual( 'lowest' );
				expect( widgets[ 1 ].slug ).toEqual( 'mediumOne' );
				expect( widgets[ 2 ].slug ).toEqual( 'mediumTwo' );
				expect( widgets[ 3 ].slug ).toEqual( 'highest' );
			} );

			it( 'should not return widgets that have been assigned but not registered', () => {
				registry.dispatch( STORE_NAME ).assignWidgetArea( 'dashboard-header', 'dashboard' );
				registry.dispatch( STORE_NAME ).assignWidget( 'PageViews', 'dashboard-header' );

				const widgets = registry.select( STORE_NAME ).getWidgets( 'dashboard-header' );

				expect( widgets ).toHaveLength( 0 );
			} );

			it( 'should return widgets that have been assigned before they were registered, once they have been registered', () => {
				const PageViews = () => {
					return ( <div>Only one person viewed your page, and it was you :-(</div> );
				};
				registry.dispatch( STORE_NAME ).assignWidgetArea( 'dashboard-header', 'dashboard' );
				registry.dispatch( STORE_NAME ).assignWidget( 'PageViews', 'dashboard-header' );
				registry.dispatch( STORE_NAME ).registerWidget( 'PageViews', {
					component: PageViews,
				} );

				const widgets = registry.select( STORE_NAME ).getWidgets( 'dashboard-header' );

				expect( widgets ).toHaveLength( 1 );
			} );
		} );

		describe( 'isWidgetRegistered', () => {
			it( 'returns true if the widget is registered', () => {
				registry.dispatch( STORE_NAME ).registerWidget( 'TestWidget', {
					component: () => {
						return ( <div>Hello test.</div> );
					},
				} );

				expect( registry.select( STORE_NAME ).isWidgetRegistered( 'TestWidget' ) ).toEqual( true );
			} );

			it( 'returns false if the widget is not registered', () => {
				expect( registry.select( STORE_NAME ).isWidgetAreaRegistered( 'NotRealWidget' ) ).toEqual( false );
			} );
		} );

		describe( 'getWidget', () => {
			it( 'returns a widget if one exists', () => {
				registry.dispatch( STORE_NAME ).registerWidget( 'TestWidget', {
					component: () => {
						return ( <div>Hello test.</div> );
					},
				} );

				expect( registry.select( STORE_NAME ).getWidget( 'TestWidget' ) ).toMatchSnapshot();
			} );

			it( 'returns null if the widget is not registered', () => {
				expect( registry.select( STORE_NAME ).getWidget( 'NotRealWidget' ) ).toEqual( null );
			} );
		} );
	} );
} );
