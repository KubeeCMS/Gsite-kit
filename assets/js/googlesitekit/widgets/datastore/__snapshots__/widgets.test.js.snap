// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`core/widgets Widgets actions registerWidget places the widget component in a variable external to the store, with a key from the store 1`] = `
<div>
  Hello 
  world
  !
</div>
`;

exports[`core/widgets Widgets selectors getWidget returns a widget if one exists 1`] = `
Object {
  "component": [Function],
  "priority": 10,
  "slug": "TestWidget",
  "width": "quarter",
  "wrapWidget": true,
}
`;
